wxRadioButton, navigation among siblings

csomor opened the pull request:
The existing functionality is exposed via base class methods

csomor committed:
First implementation

csomor committed:
Removing const for first version

csomor committed:
Going private with the implementations

Although they’re still exported of course …

vadz commented on include/wx/msw/radiobut.h:
```diff
@@ -13,7 +13,7 @@
 #include "wx/msw/ownerdrawnbutton.h"
 
-class WXDLLIMPEXP_CORE wxRadioButton : public wxMSWOwnerDrawnButton<wxControl>
+class WXDLLIMPEXP_CORE wxRadioButton : public public wxRadioButtonBase<wxMSWOwnerDrawnButton<wxControl>>
```
```suggestion
class WXDLLIMPEXP_CORE wxRadioButton : public wxRadioButtonBase< wxMSWOwnerDrawnButton<wxControl> >
```
(there are 2 changes here: fix the extra `public` and add a space for C++98 compatibility)

csomor replied:
> Thanks, no real comments, but we need to fix the build, of course, and we also need to document the new functions in `interface/wx/radiobut.h`. Ideal would be to also add some unit tests for them, but I guess we can live without them as this is not really new code...

Thanks, yes, I overlooked the Win builds ...

vadz reviewed (commented):
Thanks, no real comments, but we need to fix the build, of course, and we also need to document the new functions in `interface/wx/radiobut.h`. Ideal would be to also add some unit tests for them, but I guess we can live without them as this is not really new code...

csomor committed:
Update include/wx/msw/radiobut.h

Co-authored-by: VZ <vz-github@zeitlins.org>

csomor committed:
Adding docs

csomor commented:
@vadz, I've added simple one-liners, do we need more ?

vadz commented:
Sorry, I don't see the updates, haven't you actually pushed them yet?

In any case, there is not much to say about these methods, but I'd at least mention the behaviour for `wxRB_SINGLE` case.

csomor committed:
Added info about wxRB_SINGLE

csomor commented:
sorry, of course, done now, and I've added wxRB_SINGLE

vadz committed:
Improve radio button groups documentation

Explain the role of wxRB_SINGLE better, it's useful not only for
avoiding wxMSW bugs (which, besides, shouldn't exist any more).

vadz commented:
Sorry, I don't think the documentation of `wxRB_SINGLE` case is correct -- I asked about it because I wasn't sure, but looking at it now, I see that it actually returns the button itself from `Get{First,Last}` in this case. I'll fix this in a moment.

vadz committed:
Mark radio button group navigation functions as new in 3.1.5

Just add the missing "@since" lines.

vadz committed:
Just use wxPrivate namespace

Writing "wxPrivate::wxFoo()" is a bit ugly, so just add a using
directive for the namespace, it's not a problem to do it in this file.

vadz committed:
Remove CppUnit boilerplate from wxRadioButton unit test

Replace CppUnit test case class with a simple fixture and use CHECK()
instead of CPPUNIT_ASSERT_XXX().

No real changes.

csomor commented:
yes, your right, sorry, I overlooked this completely

vadz committed:
Use wxScopedPtr<> instead of manual delete in wxRadioButton test

Make the test code safer and ensure that no controls created here remain
alive after the test end.

vadz commented:
Trying to add a test for these functions I've also discovered that they currently can't be used (the test doesn't link) under Linux, so I'm going to fix this too, but it will take a bit longer...

vadz committed:
Don't DLL-export wxRadioButtonBase template class

This is unnecessary, this class only has inline functions.

vadz committed:
Compile radio button group navigation functions on all platforms

Previously they were only used, and compiled, on the platforms without
wxHAS_NATIVE_TAB_TRAVERSAL, i.e. were not compiled at all in wxGTK, but
we now need them everywhere as they're used to implement public API.

vadz committed:
Make radio button navigation functions const

This requires adding a couple of const_cast<>s in their implementation
in order to still allow them returning non-const wxRadioButton pointers,
but this seems preferable to not being able to call them on a const
wxRadioButton in the first place.

vadz committed:
Add unit tests for radio button group navigation functions

Extend the existing "group" and "single" tests to check these functions
too.

This at least verifies that these functions can be used.

vadz committed:
Make it simpler to write useful tests comparing windows

Allow creating wxWindowPtr from wxScopedPtr<> too, to avoid having to
use .get() in the test code, and add CHECK_SAME_WINDOW() macro which
gives more useful information about the windows in case of failure.

vadz committed:
Improve radio button navigation functions documentation

Correct the previously wrong described behaviour for wxRB_SINGLE buttons
and also mention that Get{First,Last}InGroup() never return NULL.

vadz committed:
Add wxRadioButtonBase::{Set,Get}Value()

Now that we do have wxRadioButtonBase class, declare wxRadioButton API
methods as pure virtual in it, to force the derived classes to implement
them.

Also remove the outdated comment saying that there is no base class for
wxRadioButtons in different ports, this is not true any longer.

vadz commented:
I couldn't prevent myself from making a few more improvements, but I'm finally done here, I think, so if the CI builds pass and if you don't have any objections, I'm going to merge this soon.

csomor commented:
thanks a lot

vadz committed:
Make wxRadioButtonBase a plain class, not template

Use a dirty hack to accommodate wxUniv by deriving wxRadioButtonBase
from wxCheckBox, rather than wxControl, there. This is not pretty, but
should be addressed by refactoring wxUniv code and in the meanwhile all
the other ports don't have to bother with using a template class
unnecessarily.

vadz committed:
Move radio group navigation functions to wxRadioButtonBase

Now that this class is not a template any longer, we can have the code
for radio button group navigation directly in it, without making it
inline, so move the existing functions bodies into the new methods and
remove the old functions entirely.

No real changes, this is just a refactoring.

vadz commented:
Sorry, no, after looking at all this once again, I've realized that we could avoid having a template `wxRadioButtonBase` rather easily, after all, so I've decided to simplify the code, especially because it also allows to put the code directly in `wxRadioButtonBase` methods instead of using extra functions.

But **this** one is really the final version, promised.

vadz committed:
Don't compile wxGetSelectedButtonInGroup() if it's unused

Now that this function is static, not using it results in warnings when
building the ports not using it (e.g. wxX11).
